build_functions/CMakeLists.txt

```diff
@@ -47,6 +47,203 @@ function(target_link_libraries_with_arg_prefix arg_prefix whatIsBuilding lib)
+               FOLDER ${solution_folder})
+    
+    if (TARGET azure_c_logging)
+        target_link_libraries(${whatIsBuilding}_dll azure_c_logging)
+    endif()
+    if (TARGET umock_c)
+        target_link_libraries(${whatIsBuilding}_dll umock_c)
+    endif()
+    if (TARGET ctest)
+        target_link_libraries(${whatIsBuilding}_dll ctest)
+    endif()
+    if (TARGET testrunnerswitcher)
+        target_link_libraries(${whatIsBuilding}_dll testrunnerswitcher)
+    endif()
+    
+endfunction()
+
+function(build_exe whatIsBuilding solution_folder)
+    
+    #lazily build _lib which is needed by both exe and dll
+    if(NOT TARGET ${whatIsBuilding}_lib)
+        build_lib(${whatIsBuilding} ${solution_folder} ${ARGN})
+    endif()
+    
+    #this is the exe run by ctest (or directly from visual studio)
+    add_executable(${whatIsBuilding}_exe
+        ${${whatIsBuilding}_test_files}
+        main.c
+    )
+    
+    #make sure to compile the test file as C!
+    set_source_files_properties(${whatIsBuilding}_test_files PROPERTIES LANGUAGE C)
+    
+    #link with the common lib (exe an dll share the lib part)
+    target_link_libraries(${whatIsBuilding}_exe ${whatIsBuilding}_lib)
+    
+    set_target_properties(${whatIsBuilding}_exe
+               PROPERTIES
+               FOLDER ${solution_folder})
+    
+    target_compile_definitions(${whatIsBuilding}_exe PUBLIC -DUSE_CTEST)
+    target_include_directories(${whatIsBuilding}_exe PUBLIC ${sharedutil_include_directories})
+    
+    if (TARGET azure_c_logging)
+        target_link_libraries(${whatIsBuilding}_exe azure_c_logging)
+    endif()
+    if (TARGET umock_c)
+        target_link_libraries(${whatIsBuilding}_exe umock_c)
+    endif()
+    if (TARGET ctest)
+        target_link_libraries(${whatIsBuilding}_exe ctest)
+    endif()
+    if (TARGET testrunnerswitcher)
+        target_link_libraries(${whatIsBuilding}_exe testrunnerswitcher)
+    endif()
+
+    #register the exe with ctest's list of tests
+    add_test(NAME ${whatIsBuilding} COMMAND ${whatIsBuilding}_exe)
+    
+endfunction()
+
+#drop in replacement for add_subdirectory :)
+function(build_test_folder test_folder)
+
+    if(
+        (("${test_folder}" MATCHES ".*ut.*") AND ${run_unittests}) OR
+        (("${test_folder}" MATCHES ".*e2e.*") AND ${run_e2e_tests}) OR
+        (("${test_folder}" MATCHES ".*int.*") AND ${run_int_tests}) OR
+        (("${test_folder}" MATCHES ".*perf.*") AND ${run_perf_tests})
+    )
+        set(building exe)
+        add_subdirectory(${test_folder} ${test_folder}/${building})
+        
+        if ((${use_cppunittest}) AND (WIN32))
+            set(building dll)
+            add_subdirectory(${test_folder} ${test_folder}/${building})
+        endif()
+    else()
+        #message("test_folder is ${test_folder} NOT BUILDING")
+    endif()
+ 
+endfunction()
+
+#drop in replacement for build_c_tests :)
+function(build_test_artefacts whatIsBuilding use_gballoc solution_folder )
```

Comment:
>artefacts [](start = 20, length = 9)

artifacts

Reply:
Changed

---
In reply to: [455884232](https://github.com/Azure/azure-c-build-tools/pull/30#discussion_r455884232) [](ancestors = 455884232)